contrib/runners/action_chain_runner/dist_utils.py

```diff
@@ -80,12 +68,43 @@ def fetch_requirements(requirements_file_path):
     """
     links = []
     reqs = []
-    for req in parse_requirements(requirements_file_path, session=False):
-        # Note: req.url was used before 9.0.0 and req.link is used in all the recent versions
-        link = getattr(req, 'link', getattr(req, 'url', None))
-        if link:
-            links.append(str(link))
-        reqs.append(str(req.req))
+
+    def _get_link(line):
```

Comment:
I'm wondering how error-prone is that code block?
Did you take the rules from upstream or it's something else you consulted for code logic?

Reply:
I don't check pip code itself, but I compared the results (which is the only thing I care about at this point - the function output).

If you check the tests, you will see we have tests which compare the result of our version and pip version to ensure they are the same.

It's likely there are still some edge cases somewhere, but it's not a big deal at this point since for our current ``requirements.txt``, the end result is the same.

Reply:
This code did not support environment markers from pip, so I had to add (hacky) support for them in #4895.

@Kami Why can the `dist_utils.py` module not use any third party dependencies?

```python
# NOTE: This script can't rely on any 3rd party dependency so we need to use this code here
```

Instead of importing from pip directly (which we should never have been doing in the first place, IMO), *if* we can use third party dependencies, we can utilize the [`packaging` package](https://packaging.pypa.io/en/latest/requirements/). It has [an entire section](https://packaging.pypa.io/en/latest/requirements/) devoted to reading and parsing `requirements.txt` files, and that would be kept up-to-date with what pip and setuptools support, and would lower our maintenance burden of this code.